include/glow/Backends/TraceEvents.h

```diff
@@ -121,22 +127,68 @@ class TraceContext {
   /// on different contexts.
   int traceThread_{0};
 
+  /// Lock around traceEvents_.
+  std::mutex lock_;
+
 public:
   TraceContext(TraceLevel level, int thread)
       : traceLevel_(level), traceThread_(thread) {}
+
   /// \returns TraceEvents for the last run.
   std::vector<TraceEvent> &getTraceEvents() { return traceEvents_; }
 
+  /// \returns the integer thread id used for logged events in this context.
   int getTraceThread() const { return traceThread_; }
+
+  /// Sets the integer thread id used for logged events in this context.
   void setTraceThread(int tid) { traceThread_ = tid; }
 
+  /// \returns the level of verbosity allowed for TraceEvents.
   TraceLevel getTraceLevel() { return traceLevel_; }
+
+  /// Sets the level of verbosity for TraceEvents.
   void setTraceLevel(TraceLevel level) { traceLevel_ = level; }
 
-  void logTraceEvent(llvm::StringRef name, llvm::StringRef type = "i",
-                     std::map<std::string, std::string> args = {});
+  /// Logs a new TraceEvent at the current time with the given \p name, \p type
+  /// and optionally additional attributes.
+  void
+  logTraceEvent(llvm::StringRef name,
+                llvm::StringRef type = TraceEvent::InstantType,
+                std::map<std::string, std::string> additionalAttributes = {});
+
+  // Logs a new TraceEvent at the provided \p timestamp, with the given \p name,
+  // \p type and optionally additional attributes.
+  void
+  logTraceEvent(llvm::StringRef name, llvm::StringRef type, uint64_t timestamp,
+                std::map<std::string, std::string> additionalAttributes = {});
 };
 
+/// Macros to make logging trace events easier.
```

Comment:
Consider adding something like `These macros predicate the logging of a `TraceEvent` on a validity of the given `TracingContext`.